Resource `coderd_template` setting port share level issues

When creating a `coderd_template` for the first time, it looks like it fails with the `max_port_share_level` attribute set to "owner" even though this should be the default. The template creates fine, but then the following error occurs:
```
│ Error: Client Error
│ 
│   with module.k8s-devcontainer.coderd_template.template,
│   on modules/coderd-template/main.tf line 21, in resource "coderd_template" "template":
│   21: resource "coderd_template" "template" {
│ 
│ Failed to set max port share level via update: template metadata not modified
```

If I create a template without setting max_port_share_level initially then change it, things seem to work out fine.

I set TF_LOGS to TRACE to see if I can find anything interesting, but it looks like things are actually running fine (aside from the request returning ERROR, but it just returns the same issue). I'm wondering if the response ("template metadata not modified") is actually returning an incorrect return code since there shouldn't be any changes to the template:
```
2025-02-24T11:50:05.892-0800 [TRACE] provider.terraform-provider-coderd_v0.0.9: Received downstream response: tf_resource_type=coderd_template tf_rpc=ApplyResourceChange @module=sdk.proto tf_req_duration_ms=13552 tf_proto_version=6.7 tf_provider_addr=registry.terraform.io/coder/coderd tf_req_id=aef2d974-79f1-8e45-1f96-925b53e113a8 @caller=github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov6/internal/tf6serverlogging/downstream_request.go:42 diagnostic_error_count=1 diagnostic_warning_count=0 timestamp=2025-02-24T11:50:05.892-0800
2025-02-24T11:50:05.892-0800 [ERROR] provider.terraform-provider-coderd_v0.0.9: Response contains error diagnostic: @module=sdk.proto diagnostic_summary="Client Error" tf_req_id=aef2d974-79f1-8e45-1f96-925b53e113a8 tf_resource_type=coderd_template @caller=github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov6/internal/diag/diagnostics.go:58 diagnostic_severity=ERROR tf_proto_version=6.7 tf_provider_addr=registry.terraform.io/coder/coderd tf_rpc=ApplyResourceChange diagnostic_detail="Failed to set max port share level via update: template metadata not modified" timestamp=2025-02-24T11:50:05.892-0800
2025-02-24T11:50:05.892-0800 [TRACE] provider.terraform-provider-coderd_v0.0.9: Served request: @module=sdk.proto tf_provider_addr=registry.terraform.io/coder/coderd tf_resource_type=coderd_template tf_proto_version=6.7 tf_req_id=aef2d974-79f1-8e45-1f96-925b53e113a8 tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov6/tf6server/server.go:878 timestamp=2025-02-24T11:50:05.892-0800
2025-02-24T11:50:05.893-0800 [TRACE] maybeTainted: module.k8s-devcontainer.coderd_template.template encountered an error during creation, so it is now marked as tainted
2025-02-24T11:50:05.893-0800 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/coder/coderd" is in the global cache
2025-02-24T11:50:05.893-0800 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for module.k8s-devcontainer.coderd_template.template
2025-02-24T11:50:05.893-0800 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: removing state object for module.k8s-devcontainer.coderd_template.template
2025-02-24T11:50:05.893-0800 [TRACE] evalApplyProvisioners: module.k8s-devcontainer.coderd_template.template is tainted, so skipping provisioning
2025-02-24T11:50:05.893-0800 [TRACE] maybeTainted: module.k8s-devcontainer.coderd_template.template was already tainted, so nothing to do
2025-02-24T11:50:05.893-0800 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/coder/coderd" is in the global cache
2025-02-24T11:50:05.893-0800 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for module.k8s-devcontainer.coderd_template.template
2025-02-24T11:50:05.893-0800 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: removing state object for module.k8s-devcontainer.coderd_template.template
2025-02-24T11:50:05.894-0800 [TRACE] statemgr.Filesystem: not making a backup, because the new snapshot is identical to the old
2025-02-24T11:50:05.894-0800 [TRACE] statemgr.Filesystem: no state changes since last snapshot
2025-02-24T11:50:05.894-0800 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
2025-02-24T11:50:05.907-0800 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2025-02-24T11:50:05.907-0800 [ERROR] vertex "module.k8s-devcontainer.coderd_template.template" error: Client Error
2025-02-24T11:50:05.907-0800 [TRACE] vertex "module.k8s-devcontainer.coderd_template.template": visit complete, with errors
2025-02-24T11:50:05.907-0800 [TRACE] dag/walk: upstream of "module.k8s-devcontainer (close)" errored, so skipping
2025-02-24T11:50:05.907-0800 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/coder/coderd\"] (close)" errored, so skipping
2025-02-24T11:50:05.907-0800 [TRACE] dag/walk: upstream of "root" errored, so skipping
2025-02-24T11:50:05.907-0800 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/coder/coderd" is in the global cache
2025-02-24T11:50:05.907-0800 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/hashicorp/external" is in the global cache
2025-02-24T11:50:05.907-0800 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/hashicorp/archive" is in the global cache
2025-02-24T11:50:05.908-0800 [TRACE] statemgr.Filesystem: not making a backup, because the new snapshot is identical to the old
2025-02-24T11:50:05.908-0800 [TRACE] statemgr.Filesystem: no state changes since last snapshot
2025-02-24T11:50:05.908-0800 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
```

--- a/internal/provider/template_resource.go
+++ b/internal/provider/template_resource.go
@@ -573,6 +573,8 @@ func (r *TemplateResource) Create(ctx context.Context, req resource.CreateReques
 	// deployment running `v2.15.0` or later.
 	if data.MaxPortShareLevel.IsUnknown() {
 		data.MaxPortShareLevel = types.StringValue(string(templateResp.MaxPortShareLevel))
+	} else if data.MaxPortShareLevel.ValueString() == string(templateResp.MaxPortShareLevel) {
+		tflog.Info(ctx, "max port share level set to default, not updating")
 	} else {
 		mpslReq := data.toUpdateRequest(ctx, &resp.Diagnostics)
 		if resp.Diagnostics.HasError() {

--- a/internal/provider/template_resource_test.go
+++ b/internal/provider/template_resource_test.go
@@ -657,129 +657,160 @@ func TestAccTemplateResourceEnterprise(t *testing.T) {
 	err = cp.Copy("../../integration/template-test/example-template", exTemplateOne)
 	require.NoError(t, err)
 
-	cfg1 := testAccTemplateResourceConfig{
-		URL:   client.URL.String(),
-		Token: client.SessionToken(),
-		Name:  ptr.Ref("example-template"),
-		Versions: []testAccTemplateVersionConfig{
-			{
-				// Auto-generated version name
-				Directory: &exTemplateOne,
-				Active:    ptr.Ref(true),
-			},
-		},
-		ACL: testAccTemplateACLConfig{
-			GroupACL: []testAccTemplateKeyValueConfig{
+	t.Run("BasicUsage", func(t *testing.T) {
+		cfg1 := testAccTemplateResourceConfig{
+			URL:   client.URL.String(),
+			Token: client.SessionToken(),
+			Name:  ptr.Ref("example-template"),
+			Versions: []testAccTemplateVersionConfig{
 				{
-					Key:   ptr.Ref(firstUser.OrganizationIDs[0].String()),
-					Value: ptr.Ref("use"),
-				},
-				{
-					Key:   ptr.Ref(group.ID.String()),
-					Value: ptr.Ref("admin"),
+					// Auto-generated version name
+					Directory: &exTemplateOne,
+					Active:    ptr.Ref(true),
 				},
 			},
-			UserACL: []testAccTemplateKeyValueConfig{
-				{
-					Key:   ptr.Ref(firstUser.ID.String()),
-					Value: ptr.Ref("admin"),
-				},
-			},
-		},
-	}
-
-	cfg2 := cfg1
-	cfg2.ACL.GroupACL = slices.Clone(cfg2.ACL.GroupACL[1:])
-	cfg2.MaxPortShareLevel = ptr.Ref("owner")
-
-	cfg3 := cfg2
-	cfg3.ACL.null = true
-	cfg3.MaxPortShareLevel = ptr.Ref("public")
-
-	cfg4 := cfg3
-	cfg4.AllowUserAutostart = ptr.Ref(false)
-	cfg4.AutostopRequirement = testAccAutostopRequirementConfig{
-		DaysOfWeek: ptr.Ref([]string{"monday", "tuesday"}),
-		Weeks:      ptr.Ref(int64(2)),
-	}
-
-	resource.Test(t, resource.TestCase{
-		PreCheck:                 func() { testAccPreCheck(t) },
-		IsUnitTest:               true,
-		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
-		Steps: []resource.TestStep{
-			{
-				Config: cfg1.String(t),
-				Check: resource.ComposeAggregateTestCheckFunc(
-					resource.TestCheckResourceAttr("coderd_template.test", "max_port_share_level", "owner"),
-					resource.TestCheckResourceAttr("coderd_template.test", "acl.groups.#", "2"),
-					resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.groups.*", map[string]*regexp.Regexp{
-						"id":   regexp.MustCompile(firstUser.OrganizationIDs[0].String()),
-						"role": regexp.MustCompile("^use$"),
-					}),
-					resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.groups.*", map[string]*regexp.Regexp{
-						"id":   regexp.MustCompile(group.ID.String()),
-						"role": regexp.MustCompile("^admin$"),
-					}),
-					resource.TestCheckResourceAttr("coderd_template.test", "acl.users.#", "1"),
-					resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.users.*", map[string]*regexp.Regexp{
-						"id":   regexp.MustCompile(firstUser.ID.String()),
-						"role": regexp.MustCompile("^admin$"),
-					}),
-				),
-			},
-			{
-				Config: cfg2.String(t),
-				Check: resource.ComposeAggregateTestCheckFunc(
-					resource.TestCheckResourceAttr("coderd_template.test", "max_port_share_level", "owner"),
-					resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.users.*", map[string]*regexp.Regexp{
-						"id":   regexp.MustCompile(firstUser.ID.String()),
-						"role": regexp.MustCompile("^admin$"),
-					}),
-				),
-			},
-			{
-				Config: cfg3.String(t),
-				Check: resource.ComposeAggregateTestCheckFunc(
-					resource.TestCheckResourceAttr("coderd_template.test", "max_port_share_level", "public"),
-					resource.TestCheckNoResourceAttr("coderd_template.test", "acl"),
-					func(s *terraform.State) error {
-						templates, err := client.Templates(ctx, codersdk.TemplateFilter{})
-						if err != nil {
-							return err
-						}
-						if len(templates) != 1 {
-							return fmt.Errorf("expected 1 template, got %d", len(templates))
-						}
-						acl, err := client.TemplateACL(ctx, templates[0].ID)
-						if err != nil {
-							return err
-						}
-						if len(acl.Groups) != 1 {
-							return fmt.Errorf("expected 1 group ACL, got %d", len(acl.Groups))
-						}
-						if acl.Groups[0].Role != "admin" && acl.Groups[0].ID != group.ID {
-							return fmt.Errorf("expected group ACL to be 'use' for %s, got %s", firstUser.OrganizationIDs[0].String(), acl.Groups[0].Role)
-						}
-						if len(acl.Users) != 1 {
-							return fmt.Errorf("expected 1 user ACL, got %d", len(acl.Users))
-						}
-						if acl.Users[0].Role != "admin" && acl.Users[0].ID != firstUser.ID {
-							return fmt.Errorf("expected user ACL to be 'admin' for %s, got %s", firstUser.ID.String(), acl.Users[0].Role)
-						}
-						return nil
+			ACL: testAccTemplateACLConfig{
+				GroupACL: []testAccTemplateKeyValueConfig{
+					{
+						Key:   ptr.Ref(firstUser.OrganizationIDs[0].String()),
+						Value: ptr.Ref("use"),
 					},
-				),
+					{
+						Key:   ptr.Ref(group.ID.String()),
+						Value: ptr.Ref("admin"),
+					},
+				},
+				UserACL: []testAccTemplateKeyValueConfig{
+					{
+						Key:   ptr.Ref(firstUser.ID.String()),
+						Value: ptr.Ref("admin"),
+					},
+				},
 			},
-			{
-				Config: cfg4.String(t),
-				Check: resource.ComposeAggregateTestCheckFunc(
-					resource.TestCheckResourceAttr("coderd_template.test", "allow_user_auto_start", "false"),
-					resource.TestCheckResourceAttr("coderd_template.test", "auto_stop_requirement.days_of_week.#", "2"),
-					resource.TestCheckResourceAttr("coderd_template.test", "auto_stop_requirement.weeks", "2"),
-				),
+		}
+
+		cfg2 := cfg1
+		cfg2.ACL.GroupACL = slices.Clone(cfg2.ACL.GroupACL[1:])
+		cfg2.MaxPortShareLevel = ptr.Ref("owner")
+
+		cfg3 := cfg2
+		cfg3.ACL.null = true
+		cfg3.MaxPortShareLevel = ptr.Ref("public")
+
+		cfg4 := cfg3
+		cfg4.AllowUserAutostart = ptr.Ref(false)
+		cfg4.AutostopRequirement = testAccAutostopRequirementConfig{
+			DaysOfWeek: ptr.Ref([]string{"monday", "tuesday"}),
+			Weeks:      ptr.Ref(int64(2)),
+		}
+
+		resource.Test(t, resource.TestCase{
+			PreCheck:                 func() { testAccPreCheck(t) },
+			IsUnitTest:               true,
+			ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
+			Steps: []resource.TestStep{
+				{
+					Config: cfg1.String(t),
+					Check: resource.ComposeAggregateTestCheckFunc(
+						resource.TestCheckResourceAttr("coderd_template.test", "max_port_share_level", "owner"),
+						resource.TestCheckResourceAttr("coderd_template.test", "acl.groups.#", "2"),
+						resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.groups.*", map[string]*regexp.Regexp{
+							"id":   regexp.MustCompile(firstUser.OrganizationIDs[0].String()),
+							"role": regexp.MustCompile("^use$"),
+						}),
+						resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.groups.*", map[string]*regexp.Regexp{
+							"id":   regexp.MustCompile(group.ID.String()),
+							"role": regexp.MustCompile("^admin$"),
+						}),
+						resource.TestCheckResourceAttr("coderd_template.test", "acl.users.#", "1"),
+						resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.users.*", map[string]*regexp.Regexp{
+							"id":   regexp.MustCompile(firstUser.ID.String()),
+							"role": regexp.MustCompile("^admin$"),
+						}),
+					),
+				},
+				{
+					Config: cfg2.String(t),
+					Check: resource.ComposeAggregateTestCheckFunc(
+						resource.TestCheckResourceAttr("coderd_template.test", "max_port_share_level", "owner"),
+						resource.TestMatchTypeSetElemNestedAttrs("coderd_template.test", "acl.users.*", map[string]*regexp.Regexp{
+							"id":   regexp.MustCompile(firstUser.ID.String()),
+							"role": regexp.MustCompile("^admin$"),
+						}),
+					),
+				},
+				{
+					Config: cfg3.String(t),
+					Check: resource.ComposeAggregateTestCheckFunc(
+						resource.TestCheckResourceAttr("coderd_template.test", "max_port_share_level", "public"),
+						resource.TestCheckNoResourceAttr("coderd_template.test", "acl"),
+						func(s *terraform.State) error {
+							templates, err := client.Templates(ctx, codersdk.TemplateFilter{})
+							if err != nil {
+								return err
+							}
+							if len(templates) != 1 {
+								return fmt.Errorf("expected 1 template, got %d", len(templates))
+							}
+							acl, err := client.TemplateACL(ctx, templates[0].ID)
+							if err != nil {
+								return err
+							}
+							if len(acl.Groups) != 1 {
+								return fmt.Errorf("expected 1 group ACL, got %d", len(acl.Groups))
+							}
+							if acl.Groups[0].Role != "admin" && acl.Groups[0].ID != group.ID {
+								return fmt.Errorf("expected group ACL to be 'use' for %s, got %s", firstUser.OrganizationIDs[0].String(), acl.Groups[0].Role)
+							}
+							if len(acl.Users) != 1 {
+								return fmt.Errorf("expected 1 user ACL, got %d", len(acl.Users))
+							}
+							if acl.Users[0].Role != "admin" && acl.Users[0].ID != firstUser.ID {
+								return fmt.Errorf("expected user ACL to be 'admin' for %s, got %s", firstUser.ID.String(), acl.Users[0].Role)
+							}
+							return nil
+						},
+					),
+				},
+				{
+					Config: cfg4.String(t),
+					Check: resource.ComposeAggregateTestCheckFunc(
+						resource.TestCheckResourceAttr("coderd_template.test", "allow_user_auto_start", "false"),
+						resource.TestCheckResourceAttr("coderd_template.test", "auto_stop_requirement.days_of_week.#", "2"),
+						resource.TestCheckResourceAttr("coderd_template.test", "auto_stop_requirement.weeks", "2"),
+					),
+				},
 			},
-		},
+		})
+	})
+
+	// Verifies that when `max_port_share_level` is set to to the default value,
+	// an update request that would return HTTP Not Modified is not sent.
+	t.Run("DefaultMaxPortShareLevel", func(t *testing.T) {
+		cfg1 := testAccTemplateResourceConfig{
+			URL:   client.URL.String(),
+			Token: client.SessionToken(),
+			Name:  ptr.Ref("example-template"),
+			Versions: []testAccTemplateVersionConfig{
+				{
+					Directory: &exTemplateOne,
+					Active:    ptr.Ref(true),
+				},
+			},
+			MaxPortShareLevel: ptr.Ref("owner"),
+		}
+
+		resource.Test(t, resource.TestCase{
+			PreCheck:                 func() { testAccPreCheck(t) },
+			IsUnitTest:               true,
+			ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
+			Steps: []resource.TestStep{
+				{
+					Config: cfg1.String(t),
+					Check:  resource.TestCheckResourceAttr("coderd_template.test", "max_port_share_level", "owner"),
+				},
+			},
+		})
 	})
 }
 

